elegant/edge_detection.py

```diff
@@ -0,0 +1,203 @@
+    image_down = pyramid.pyr_down(image, downscale=2)
+
+    #get the gradient
+    gradient = ndimage.filters.gaussian_gradient_magnitude(image_down, 1)
+    top_ten = np.percentile(gradient,  61)
+    gradient = sigmoid(gradient, gradient.min(), top_ten, gradient.max(), 2)
+    gradient = gradient.max()-abs(gradient)
+
+    #penalize finding edges near the centerline or outside of the avg_width_tck
+    #since the typical worm is fatter than the centerline and not huge
+    #Need to divide by 2 because of the downsampling
+    pen_widths = (interpolate.spline_interpolate(avg_width_tck, image_down.shape[0]))
+    #pen_widths = pen_widths/2
+    distance_matrix = abs(np.subtract.outer(pen_widths, np.arange(0, image_down.shape[1])))
+    #distance_matrix = np.flip(abs(np.subtract.outer(pen_widths, np.arange(0, image_down.shape[1]))), 1)
+    penalty = alpha*(distance_matrix)
+    new_costs = gradient+penalty
+    
+    #set start and end points for the traceback
+    start = (0, int(pen_widths[0]))
+    end = (len(pen_widths)-1, int(pen_widths[-1]))
+    #start = (0, int((image_down.shape[1]-1)-pen_widths[0]))
+    #end = (len(pen_widths)-1, int((image_down.shape[1]-1)-pen_widths[-1]))
+    #start = (0,0)
+    #end = (len(pen_widths)-1, 0)
+
+    #begin edge detection
+    offsets= [(1,-1),(1,0),(1,1)]
+    mcp = Smooth_MCP(new_costs, mcp_alpha, offsets=offsets)
+    mcp.find_costs([start], [end])
+    route = mcp.traceback(end)
+
+    x,y = np.transpose(route)
+    #print("x: ", len(x), "y:", len(y))
+    #print(route)
+    #y = image_down.shape[1] - y
+    return (x*2, y*2) #multiply by 2 to account for downsampling
+
+def sigmoid(gradient, min, mid, max, growth_rate):
+    '''Apply the sigmoid function to the gradient.
+
+    Parameters:
+        gradient: array of the gradient of the image
+        min: lower asymptote of the sigmoid function
+        mid: midpoint of the sigmoid function (ie. the point that is halfway between
+            the lower and upper asymptotes)
+        max: upper asymptote
+        growth_rate: growth rate of the sigmoid function
+    '''
+    return min+((max-min)/(1+np.exp(-(growth_rate)*(gradient-mid))))
+
+def scale_image(image):
+    mode = np.bincount(image.flat)[1:].argmax()+1
+    bf = image.astype(np.float32)
+    bf -= 200
+    bf *= (24000-200) / (mode-200)
+    bf8 = colorize.scale(bf, min=600, max=26000, gamma=0.72, output_max=255).astype(np.float32)
+    return bf8
+
+class Smooth_MCP(graph.MCP_Flexible):
+    """Custom MCP class to weight different route possibilities.
+    Penalize sharp changes to make the end widths a little smoother
+    """
+    def __init__(self, costs, alpha, offsets=None):
+        graph.MCP_Flexible.__init__(self, costs, offsets=offsets)
+        self.alpha = alpha
+
+    def travel_cost(self, old_cost, new_cost, offset_length):
+        """Override method to smooth out the traceback
+        """
+
+        return self.alpha*(new_cost + abs(offset_length)) 
```

Comment:
The weight parameter `alpha` should control the relative cost of a horizontal vs. diagonal move. However, in the above code, it actually does nothing. Because horizontal moves (with `offset_length` of 1) and diagonal moves (with `offset_length` of 1.41...) both have costs multiplied by `alpha`, the relative penalty for a diagonal move compared to a horizontal move is always just 1.41 -- regardless of what `alpha` is, so the same path will always be found.

Also, because `offset_length` is added to the cost directly, it will have a relatively large effect when the values in the costs array are < 1 (e.g. the `offset_length` term will dominate), and have a relatively small effect if the values in the costs array are >> 1. The `alpha` parameter can't help tune this for different cost arrays.

All in all, you'll probably want something more like:
```
penalty = 1 if offset_length == 1 else self.alpha
return new_cost * penalty
```
or
```
penalty = 0 if offset_length == 1 else self.alpha
return new_cost + penalty
```
(I'm not really sure if it ought to be an additive or multiplicative penalty! Might take a bit of testing to see what seems to give the smoothest control between super-wiggly paths and super-straight ones.)

Also, when you write code like this, it's important to test in isolation first to make sure that it seems to be doing the expected thing. E.g. test with a few different values for `alpha` and make sure that with higher `alpha` you get smoother and smoother paths. The problem above would have been obvious from some testing in this vein.

Last, note that `offset_length` is probably not ever negative, no? So no need for an `abs` (though we should probably check in the skimage code to ensure that it is indeed always positive...).